Improve attribute parsing from monsters.txt

Veracity0 opened the pull request:
- I rationalized how we parse attributes from monsters.txt. Rather than having 36 different attributes and calling a MonsterData constructor with 28 parameters, parse the parameters into 1 data structure and call a constructor with 4 parameters.

The constructor still needs to extract all the attributes out and set the expected fields, but adding a new attribute now no longer requires adding a parameter to the constructor: add parsing in MonsterData, add a field to the MonsterData class and set it in the constructor, and, as before, use that field as needed. Perhaps a field in the Monster Proxy Record. And so on.

- I normalized the order of the fields in monsters.txt: Atk, Def, HP, Init, ... or something.
- I added Article, Group, Elemental Resistance as attribute, complete with Monster Proxy Record field.
- I normalize string you get from the Monster.attributes proxy record field
- I added "test fix-monster-attributes" to read monsters.txt line by line and replace the attributes (field 4) with the normalized version.
- I ran it on monsters.txt and found 998 (!) monsters that needed to be updated. This is included.
- I made Phys: and Elem: allow Expressions, rather than just ints
- I wrote a test suite for the string -> attribute map and attribute map -> string methods.
- I wrote a test showing that the MonsterData constructor parses an attribute string.

Coming in a follow-up PR:

- Improve MonsterManuelManager to extract "Article" and include it in the "attributes" string it generates for new monsters.
- Provide MonsterManuelManager with a utility to visit every page and extract the Articles for every monster, saving them in a map from monster ID to Attribute map
- Given that, it will be trivial to make "test fix-monster-articles" to add that field to every monster in Manuel.

Veracity0 commented:
Ready for review. I am deferring importing "article" from MonsterManuelManager until the next PR.

midgleyc commented on src/data/monsters.txt:
```diff
@@ -376,23 +376,23 @@ Knob Goblin Mutant	88	kg_mutant.gif	Atk: 53 Def: 47 HP: 43 Init: 70 Meat: 55 P:
 Lesser Fruit Golem	284	lfruitgol.gif	Atk: 17 Def: 15 HP: 9 Init: 50 P: construct	grapefruit (15)	grapes (15)	lemon (15)	orange (15)	strawberry (15)
-lihc	32	lich.gif	Atk: 20 Def: 18 HP: 11 Init: 60 ED: spooky EA: spooky EA: sleaze Meat: 36 P: undead	lihc eye (30)
-Little Man in the Canoe	540	canoeman.gif	Atk: 165 Def: 148 HP: 150 Init: 70 E: sleaze P: constellation	line (30)	star (30)	star (30)
+lihc	32	lich.gif	Atk: 20 Def: 18 HP: 11 Init: 60 Meat: 36 P: undead E: sleaze	lihc eye (30)
```
A lihc has four attacks, three of which are spooky elemental and one of which is sleaze. The script has observed that both EA and ED were present, and collapsed them into E, but picked the wrong EA.

To clarify, as I understand EA, ED, and E:
* ED: this monster is element X
* EA: this monster has at least one attack with damage X
* E: this monster has element X, and at least one elemental attack with damage X

Notably, EA can occur multiple times, as here.

Veracity0 replied:
In my workspace, I have restored the original monsters.txt (and the obsolete parsing of poison) so I can work on this.

Looking at the code before I started this work:
The MonsterData.java:
```
  private Element attackElement;
  private Element defenseElement;
```
There is only only attack element and one defense element.

MonsterDatabase.java
```
        } else if (option.equals("E:")) {
          if (tokens.hasMoreTokens()) {
            value = tokens.nextToken();
            Element element = MonsterDatabase.parseElement(value);
            if (element != Element.NONE) {
              attackElement = element;
              defenseElement = element;
            }
          }
          continue;
        } else if (option.equals("ED:")) {
          if (tokens.hasMoreTokens()) {
            value = tokens.nextToken();
            Element element = MonsterDatabase.parseElement(value);
            if (element != Element.NONE) {
              defenseElement = element;
            }
          }
          continue;
        } else if (option.equals("EA:")) {
          if (tokens.hasMoreTokens()) {
            value = tokens.nextToken();
            Element element = MonsterDatabase.parseElement(value);
            if (element != Element.NONE) {
              attackElement = element;
            }
          }
          continue;
        }
```
I think your understanding of the attributes is correct.
Notice that since we do not currently support multiple elemental attacks, we simply store the last one found.
So, yeah - my collapsing of EA+ED to E should check that both EA and ED are the same element.
Unfortunately, since we do not store additional EA, I can't get both without doing something about "we only store one EA".

I'll ponder this. I wonder how many more monsters there are with more than one EA?
I can debug that by logging when we parse the strings from monsters.txt. :)

midgleyc replied:
Some monsters, like Lord Spookyraven, can attack with any element.

Veracity0 replied:
I have fixed the handling for collapsing EA and ED into E.
I have left multiple EA attributes as is: only keep the last one listed.
However, I made it log when it sees this - only the lihc, at the moment - so we will be reminded to maybe think more about this. :)

midgleyc commented on src/data/monsters.txt:
```diff
@@ -376,23 +376,23 @@ Knob Goblin Mutant	88	kg_mutant.gif	Atk: 53 Def: 47 HP: 43 Init: 70 Meat: 55 P:
 Knob Goblin Very Mad Scientist	86	kg_verymadsci.gif	Atk: 45 Def: 40 HP: 30 Init: 70 Meat: 45 P: goblin	Knob Goblin melon baller (5)	Knob Goblin seltzer (15)	Knob Goblin superseltzer (15)	scrumptious reagent (25)	bunch of square grapes (c19)	smudged alchemical recipe (p5)	Cobb's Knob Menagerie key (c0)	Subject 37 file (c0)
-Knott Slanding	617	slanding.gif	NOCOPY ULTRARARE NOMANUEL Atk: 150 Def: 135 HP: 150 Init: 10000 E: cold Meat: 970-1342 P: beast	Dallas Dynasty Falcon Crest shield (100)
-Knott Yeti	99	yeti.gif	Atk: 105 Def: 92 HP: 90 Init: 60 E: cold Meat: 200 P: beast	yeti fur (30)
+Knott Slanding	617	slanding.gif	NOCOPY NOMANUEL ULTRARARE Atk: 150 Def: 135 HP: 150 Init: 10000 Meat: 1156 P: beast E: cold	Dallas Dynasty Falcon Crest shield (100)
```
I think meat should be left dashed in this case: it shows more clearly that these are values determined from fights, instead of from more thorough spading. I doubt the true value is 1156.

Veracity0 replied:
Interesting. Again, the old code did the collapsing and passed a single integer to the MonsterData constructor.
Perhaps I should have Meat: have either an Integer or a String and the constructor does the collapsing.
I wonder how many monsters have this?

The number which is stored is the average. ASH proxy fields "min_meat" and "max_meat" are calculated: the stored value +- 20%

Again, this is an issue for going back from Attributes to String.

Veracity0 replied:
I fixed this to calculate (and store) average meat from a range, but leave the range in the attribute.
Debug logging showed this only affects the Knott Slanding.
That's interesting, but we don't need to nag about this single monster, so I disabled the logging.

midgleyc commented:
Checking `monsters.txt` for lines of differing length (then removing those where `Poison` was added) gives:
```
263 full-length mirror	1553	bigmirror.gif	NOCOPY LUCKY Scale: 0 Cap: ? Init: 50 P: construct
263 full-length mirror	1553	bigmirror.gif	NOCOPY LUCKY Scale: 0 Cap: 10000 Init: 50 P: construct
292 Gnollish Crossdresser	19	dk_cross.gif	Atk: 11 Def: 9 HP: 5 Init: 60 E: sleaze EA: sleaze Meat: 22 P: humanoid	gnoll teeth (5)	frilly skirt (25)	maiden wig (25)	frilly shirt (c5)	Gnollish Crossdress (c0)
292 Gnollish Crossdresser	19	dk_cross.gif	Atk: 11 Def: 9 HP: 5 Init: 60 Meat: 22 P: humanoid E: sleaze	gnoll teeth (5)	frilly skirt (25)	maiden wig (25)	frilly shirt (c5)	Gnollish Crossdress (c0)
378 Knott Slanding	617	slanding.gif	NOCOPY ULTRARARE NOMANUEL Atk: 150 Def: 135 HP: 150 Init: 10000 E: cold Meat: 970-1342 P: beast	Dallas Dynasty Falcon Crest shield (100)
378 Knott Slanding	617	slanding.gif	NOCOPY NOMANUEL ULTRARARE Atk: 150 Def: 135 HP: 150 Init: 10000 Meat: 1156 P: beast E: cold	Dallas Dynasty Falcon Crest shield (100)
388 lihc	32	lich.gif	Atk: 20 Def: 18 HP: 11 Init: 60 ED: spooky EA: spooky EA: sleaze Meat: 36 P: undead	lihc eye (30)
388 lihc	32	lich.gif	Atk: 20 Def: 18 HP: 11 Init: 60 Meat: 36 P: undead E: sleaze	lihc eye (30)
570 scary pirate	307	zompirate.gif	Atk: 150 Def: 135 HP: 200 Init: 10000 E: spooky Meat: 45 P: pirate Wiki: "scary pirate (cursed)" 	cursed breeches (n0)	cursed cutlass (n0)	cursed eyepatch (n0)	folder (skull and crossbones) (c0)
570 scary pirate	307	zompirate.gif	Atk: 150 Def: 135 HP: 200 Init: 10000 Meat: 45 P: pirate E: spooky Wiki: "scary pirate (cursed)"	cursed breeches (n0)	cursed cutlass (n0)	cursed eyepatch (n0)	folder (skull and crossbones) (c0)
638 Stephen Spookyraven	1565	steven.gif	NOCOPY Scale: 5 Cap: 300 Init: 100 Meat: 500  P: undead	Stephen's secret formula (100)	Stephen's lab coat (100)
638 Stephen Spookyraven	1565	steven.gif	NOCOPY Scale: 5 Cap: 300 Init: 100 Meat: 500 P: undead	Stephen's secret formula (100)	Stephen's lab coat (100)
681 The Thing in the Basement	2164	drippything1.gif	NOMANUEL Atk: [300] Def: [300] HP: [320] Init: -10000 P: horror DRIPPY 	The Eye of the Thing in the Basement (n100)	The Fingernail of the Thing in the Basement (n100)
681 The Thing in the Basement	2164	drippything1.gif	NOMANUEL Atk: [300] Def: [300] HP: [320] Init: -10000 P: horror DRIPPY	The Eye of the Thing in the Basement (n100)	The Fingernail of the Thing in the Basement (n100)
1240 furious cow	1926	awolcow1.gif	WANDERER Scale: 0 Cap: ? Floor: ? Init: -10000 P: demon
1240 furious cow	1926	awolcow1.gif	WANDERER Scale: 0 Cap: 10000 Floor: 10 Init: -10000 P: demon
1241 furious giant cow	1929	awolcow2.gif	WANDERER Scale: 20 Cap: ? Floor: ? Init: -10000 P: demon
1241 furious giant cow	1929	awolcow2.gif	WANDERER Scale: 20 Cap: 10000 Floor: 10 Init: -10000 P: demon
1242 ungulith	1932	awolcow3.gif	WANDERER Scale: 40 Cap: ? Floor: ? Init: -10000 P: demon
1242 ungulith	1932	awolcow3.gif	WANDERER Scale: 40 Cap: 10000 Floor: 10 Init: -10000 P: demon
1244 emaciated rodeo clown	1927	awolclown1.gif	WANDERER Scale: 0 Cap: ? Floor: ? Init: -10000 P: horror
1244 emaciated rodeo clown	1927	awolclown1.gif	WANDERER Scale: 0 Cap: 10000 Floor: 10 Init: -10000 P: horror
1245 menacing rodeo clown	1930	awolclown2.gif	WANDERER Scale: 20 Cap: ? Floor: ? Init: -10000 P: horror
1245 menacing rodeo clown	1930	awolclown2.gif	WANDERER Scale: 20 Cap: 10000 Floor: 10 Init: -10000 P: horror
1246 grizzled rodeo clown	1933	awolclown3.gif	WANDERER Scale: 40 Cap: ? Floor: ? Init: -10000 P: horror
1246 grizzled rodeo clown	1933	awolclown3.gif	WANDERER Scale: 40 Cap: 10000 Floor: 10 Init: -10000 P: horror
1248 aggressive grass snake	1928	awolsnake1.gif	WANDERER Scale: 0 Cap: ? Floor: ? Init: -10000 P: beast SNAKE
1248 aggressive grass snake	1928	awolsnake1.gif	WANDERER Scale: 0 Cap: 10000 Floor: 10 Init: -10000 P: beast SNAKE
1249 prince snake	1931	awolsnake2.gif	WANDERER Scale: 20 Cap: ? Floor: ? Init: -10000 E: sleaze P: beast SNAKE	snake oil (100)	snake oil (100)
1249 prince snake	1931	awolsnake2.gif	WANDERER Scale: 20 Cap: 10000 Floor: 10 Init: -10000 P: beast SNAKE E: sleaze	snake oil (100)	snake oil (100)
1250 king snake	1934	awolsnake3.gif	WANDERER Scale: 40 Cap: ? Floor: ? Init: -10000 P: beast SNAKE	snake oil (100)	snake oil (100)	snake oil (100)
1250 king snake	1934	awolsnake3.gif	WANDERER Scale: 40 Cap: 10000 Floor: 10 Init: -10000 P: beast SNAKE	snake oil (100)	snake oil (100)	snake oil (100)
1258 Villainous Minion	2024	bond_minion1.gif,bond_minion2.gif,bond_minion3.gif,bond_minion4.gif	Scale: -2 Cap: ? Floor: ? Init: 50 P: dude
1258 Villainous Minion	2024	bond_minion1.gif,bond_minion2.gif,bond_minion3.gif,bond_minion4.gif	Scale: -2 Cap: 10000 Floor: 10 Init: 50 P: dude
1259 Villainous Henchperson	2025	bond_sidekick1.gif,bond_sidekick2.gif,bond_sidekick3.gif,bond_sidekick4.gif,bond_sidekick5.gif,bond_sidekick6.gif,bond_sidekick7.gif,bond_sidekick8.gif,bond_sidekick9.gif	NOCOPY Scale: 10 Cap: ? Floor: ? Init: -10000 P: dude
1259 Villainous Henchperson	2025	bond_sidekick1.gif,bond_sidekick2.gif,bond_sidekick3.gif,bond_sidekick4.gif,bond_sidekick5.gif,bond_sidekick6.gif,bond_sidekick7.gif,bond_sidekick8.gif,bond_sidekick9.gif	NOCOPY Scale: 10 Cap: 10000 Floor: 10 Init: -10000 P: dude
1260 Villainous Villain	2026	bond_villain1.gif,bond_villain2.gif,bond_villain3.gif,bond_villain4.gif,bond_villain5.gif,bond_villain6.gif,bond_villain7.gif	NOCOPY Scale: 11 Cap: ? Floor: ? Init: -10000 P: dude
1260 Villainous Villain	2026	bond_villain1.gif,bond_villain2.gif,bond_villain3.gif,bond_villain4.gif,bond_villain5.gif,bond_villain6.gif,bond_villain7.gif	NOCOPY Scale: 11 Cap: 10000 Floor: 10 Init: -10000 P: dude
1288 invader bullet	2137	invaderbullet.gif	WANDERER Scale: 20 Cap: ? Floor: ? Init: 10000 P: construct
1288 invader bullet	2137	invaderbullet.gif	WANDERER Scale: 20 Cap: 10000 Floor: 10 Init: 10000 P: construct
1515 Black Crayon Beast	1187	cray_beast.gif	FREE Scale: 0 Cap: ? Init: -10000 P: beast	crayon shavings (100)
1515 Black Crayon Beast	1187	cray_beast.gif	FREE Scale: 0 Cap: 10000 Init: -10000 P: beast	crayon shavings (100)
1516 Black Crayon Beetle	1191	cray_bug.gif	FREE Scale: 0 Cap: ? Init: -10000 P: bug	crayon shavings (100)
1516 Black Crayon Beetle	1191	cray_bug.gif	FREE Scale: 0 Cap: 10000 Init: -10000 P: bug	crayon shavings (100)
1517 Black Crayon Constellation	1206	cray_const.gif	FREE Scale: 0 Cap: ? Init: -10000 P: constellation	crayon shavings (100)
1517 Black Crayon Constellation	1206	cray_const.gif	FREE Scale: 0 Cap: 10000 Init: -10000 P: constellation	crayon shavings (100)
1518 Black Crayon Crimbo Elf	1201	cray_elf.gif	FREE Scale: 0 Cap: ? Init: -10000 P: elf	crayon shavings (100)
1518 Black Crayon Crimbo Elf	1201	cray_elf.gif	FREE Scale: 0 Cap: 10000 Init: -10000 P: elf	crayon shavings (100)
1519 Black Crayon Demon	1194	cray_demon.gif	FREE Scale: 0 Cap: ? Init: -10000 P: demon	crayon shavings (100)
1519 Black Crayon Demon	1194	cray_demon.gif	FREE Scale: 0 Cap: 10000 Init: -10000 P: demon	crayon shavings (100)
1520 Black Crayon Elemental	1205	cray_elemental.gif	FREE Scale: 0 Cap: ? Init: -10000 P: elemental	crayon shavings (100)
1520 Black Crayon Elemental	1205	cray_elemental.gif	FREE Scale: 0 Cap: 10000 Init: -10000 P: elemental	crayon shavings (100)
1521 Black Crayon Fish	1196	cray_fish.gif	FREE Scale: 0 Cap: ? Init: -10000 P: fish	crayon shavings (100)
1521 Black Crayon Fish	1196	cray_fish.gif	FREE Scale: 0 Cap: 10000 Init: -10000 P: fish	crayon shavings (100)
1522 Black Crayon Flower	1199	cray_plant.gif	FREE Scale: 0 Cap: ? Init: -10000 P: plant	crayon shavings (100)
1522 Black Crayon Flower	1199	cray_plant.gif	FREE Scale: 0 Cap: 10000 Init: -10000 P: plant	crayon shavings (100)
1523 Black Crayon Frat Orc	1193	cray_orc.gif	FREE Scale: 0 Cap: ? Init: -10000 P: orc	crayon shavings (100)
1523 Black Crayon Frat Orc	1193	cray_orc.gif	FREE Scale: 0 Cap: 10000 Init: -10000 P: orc	crayon shavings (100)
1524 Black Crayon Goblin	1197	cray_goblin.gif	FREE Scale: 0 Cap: ? Init: -10000 P: goblin	crayon shavings (100)
1524 Black Crayon Goblin	1197	cray_goblin.gif	FREE Scale: 0 Cap: 10000 Init: -10000 P: goblin	crayon shavings (100)
1525 Black Crayon Golem	1188	cray_construct.gif	FREE Scale: 0 Cap: ? Init: -10000 P: construct	crayon shavings (100)
1525 Black Crayon Golem	1188	cray_construct.gif	FREE Scale: 0 Cap: 10000 Init: -10000 P: construct	crayon shavings (100)
1526 Black Crayon Hippy	1192	cray_hippy.gif	FREE Scale: 0 Cap: ? Init: -10000 P: hippy	crayon shavings (100)
1526 Black Crayon Hippy	1192	cray_hippy.gif	FREE Scale: 0 Cap: 10000 Init: -10000 P: hippy	crayon shavings (100)
1527 Black Crayon Hobo	1207	cray_hobo.gif	FREE Scale: 0 Cap: ? Init: -10000 P: hobo	crayon shavings (100)
1527 Black Crayon Hobo	1207	cray_hobo.gif	FREE Scale: 0 Cap: 10000 Init: -10000 P: hobo	crayon shavings (100)
1528 Black Crayon Man	1186	cray_dude.gif	FREE Scale: 0 Cap: ? Init: -10000 P: dude	crayon shavings (100)
1528 Black Crayon Man	1186	cray_dude.gif	FREE Scale: 0 Cap: 10000 Init: -10000 P: dude	crayon shavings (100)
1529 Black Crayon Manloid	1190	cray_humanoid.gif	FREE Scale: 0 Cap: ? Init: -10000 P: humanoid	crayon shavings (100)
1529 Black Crayon Manloid	1190	cray_humanoid.gif	FREE Scale: 0 Cap: 10000 Init: -10000 P: humanoid	crayon shavings (100)
1530 Black Crayon Mer-kin	1202	cray_merkin.gif	FREE Scale: 0 Cap: ? Init: -10000 P: mer-kin	crayon shavings (100)
1530 Black Crayon Mer-kin	1202	cray_merkin.gif	FREE Scale: 0 Cap: 10000 Init: -10000 P: mer-kin	crayon shavings (100)
1531 Black Crayon Penguin	1204	cray_penguin.gif	FREE Scale: 0 Cap: ? Init: -10000 P: penguin	crayon shavings (100)
1531 Black Crayon Penguin	1204	cray_penguin.gif	FREE Scale: 0 Cap: 10000 Init: -10000 P: penguin	crayon shavings (100)
1532 Black Crayon Pirate	1198	cray_pirate.gif	FREE Scale: 0 Cap: ? Init: -10000 P: pirate	crayon shavings (100)
1532 Black Crayon Pirate	1198	cray_pirate.gif	FREE Scale: 0 Cap: 10000 Init: -10000 P: pirate	crayon shavings (100)
1533 Black Crayon Shambling Monstrosity	1195	cray_horror.gif	FREE Scale: 0 Cap: ? Init: -10000 P: horror	crayon shavings (100)
1533 Black Crayon Shambling Monstrosity	1195	cray_horror.gif	FREE Scale: 0 Cap: 10000 Init: -10000 P: horror	crayon shavings (100)
1534 Black Crayon Slime	1203	cray_slime.gif	FREE Scale: 0 Cap: ? Init: -10000 P: slime	crayon shavings (100)
1534 Black Crayon Slime	1203	cray_slime.gif	FREE Scale: 0 Cap: 10000 Init: -10000 P: slime	crayon shavings (100)
1535 Black Crayon Spiraling Shape	1200	cray_weird.gif	FREE Scale: 0 Cap: ? Init: -10000 P: weird	crayon shavings (100)
1535 Black Crayon Spiraling Shape	1200	cray_weird.gif	FREE Scale: 0 Cap: 10000 Init: -10000 P: weird	crayon shavings (100)
1536 Black Crayon Undead Thing	1189	cray_undead.gif	FREE Scale: 0 Cap: ? Init: -10000 P: undead	crayon shavings (100)
1536 Black Crayon Undead Thing	1189	cray_undead.gif	FREE Scale: 0 Cap: 10000 Init: -10000 P: undead	crayon shavings (100)
1622 gummi plesiosaur	1491	plesio.gif	NOCOPY Scale: ? Cap: ? Floor: ? Init: -10000 P: beast	small gummi fin (0)	gummi sword (0)
1622 gummi plesiosaur	1491	plesio.gif	NOCOPY Scale: 0 Cap: 10000 Floor: 10 Init: -10000 P: beast	small gummi fin (0)	gummi sword (0)
1623 lemonhead minnow	1492	lemonfish.gif	NOCOPY Scale: ? Cap: ? Floor: ? Init: -10000 P: fish	lemonhead caviar (0)	lemony scales (0)	lemony scales (0)
1623 lemonhead minnow	1492	lemonfish.gif	NOCOPY Scale: 0 Cap: 10000 Floor: 10 Init: -10000 P: fish	lemonhead caviar (0)	lemony scales (0)	lemony scales (0)
1624 school of gummi piranhas	1490	gummifish.gif	NOCOPY Scale: ? Cap: ? Floor: ? Init: -10000 P: fish	gummi fang (0)	chocolate cow bone (0)
1624 school of gummi piranhas	1490	gummifish.gif	NOCOPY Scale: 0 Cap: 10000 Floor: 10 Init: -10000 P: fish	gummi fang (0)	chocolate cow bone (0)
1625 chocolate hare	1494	chocohare.gif	NOCOPY Scale: ? Cap: ? Floor: ? Init: -10000 P: beast	chocolate rabbit's foot (0)	candy carrot (0)
1625 chocolate hare	1494	chocohare.gif	NOCOPY Scale: 0 Cap: 10000 Floor: 10 Init: -10000 P: beast	chocolate rabbit's foot (0)	candy carrot (0)
1626 chocolate-cherry prairie dog	1495	ccprairie.gif	NOCOPY Scale: ? Cap: ? Floor: ? Init: -10000 P: beast	candy mountain oyster (0)	chocolate-stained collar (0)
1626 chocolate-cherry prairie dog	1495	ccprairie.gif	NOCOPY Scale: 0 Cap: 10000 Floor: 10 Init: -10000 P: beast	candy mountain oyster (0)	chocolate-stained collar (0)
1627 Rock Pop weasel	1493	popweasel.gif	NOCOPY Scale: ? Cap: ? Floor: ? Init: -10000 P: beast	mulberry (0)	weasel stomping pants (0)
1627 Rock Pop weasel	1493	popweasel.gif	NOCOPY Scale: 0 Cap: 10000 Floor: 10 Init: -10000 P: beast	mulberry (0)	weasel stomping pants (0)
1628 candied pecan tree	1496	pecantree.gif	NOCOPY Scale: ? Cap: ? Floor: ? Init: -10000 P: plant	candied pecan (0)	candy stick (0)
1628 candied pecan tree	1496	pecantree.gif	NOCOPY Scale: 0 Cap: 10000 Floor: 10 Init: -10000 P: plant	candied pecan (0)	candy stick (0)
1629 licorice snake	1497	licosnake.gif	NOCOPY Scale: ? Cap: ? Floor: ? Init: -10000 P: beast SNAKE	anise-flavored venom (0)	licorice whip (0)
1629 licorice snake	1497	licosnake.gif	NOCOPY Scale: 0 Cap: 10000 Floor: 10 Init: -10000 P: beast SNAKE	anise-flavored venom (0)	licorice whip (0)
1630 tricksy pixie	1498	trixiepixie.gif	NOCOPY Scale: ? Cap: ? Floor: ? Init: -10000 P: humanoid	sour powder (0)	pixie axie (0)
1630 tricksy pixie	1498	trixiepixie.gif	NOCOPY Scale: 0 Cap: 10000 Floor: 10 Init: -10000 P: humanoid	sour powder (0)	pixie axie (0)
1631 fire truck	1499	firetruck.gif	NOCOPY Scale: ? Cap: ? Floor: ? Init: -10000 P: construct	fireman's lunch (0)	fire hose (0)
1631 fire truck	1499	firetruck.gif	NOCOPY Scale: 0 Cap: 10000 Floor: 10 Init: -10000 P: construct	fireman's lunch (0)	fire hose (0)
1632 ice cream truck	1500	icecreamtruck.gif	NOCOPY Scale: ? Cap: ? Floor: ? Init: -10000 P: construct	ice cream sandwich (0)	plain paper hat (0)
1632 ice cream truck	1500	icecreamtruck.gif	NOCOPY Scale: 0 Cap: 10000 Floor: 10 Init: -10000 P: construct	ice cream sandwich (0)	plain paper hat (0)
1633 monster hearse	1502	monsterhearse.gif	NOCOPY Scale: ? Cap: ? Floor: ? Init: -10000 P: construct	nachos of the night (0)	skull gearshift knob (0)
1633 monster hearse	1502	monsterhearse.gif	NOCOPY Scale: 0 Cap: 10000 Floor: 10 Init: -10000 P: construct	nachos of the night (0)	skull gearshift knob (0)
1634 sewer tanker	1501	sewertruck.gif	NOCOPY Scale: ? Cap: ? Floor: ? Init: -10000 P: construct	plumber's lunch (0)	&quot;honey&quot; dipper (0)
1634 sewer tanker	1501	sewertruck.gif	NOCOPY Scale: 0 Cap: 10000 Floor: 10 Init: -10000 P: construct	plumber's lunch (0)	&quot;honey&quot; dipper (0)
1635 sketchy van	1503	sketchyvan.gif	NOCOPY Scale: ? Cap: ? Floor: ? Init: -10000 P: construct	can of Adultwitch&trade; (0)	Sketcherz&trade; (0)
1635 sketchy van	1503	sketchyvan.gif	NOCOPY Scale: 0 Cap: 10000 Floor: 10 Init: -10000 P: construct	can of Adultwitch&trade; (0)	Sketcherz&trade; (0)
1913 Ancient Skeleton with Skin still on it	1962	skinskeleton.gif	Scale: 3 Cap: ? Floor: 20 Init: -10000 P: undead	compounded experience (c100)	time residue (c0)
1913 Ancient Skeleton with Skin still on it	1962	skinskeleton.gif	Scale: 3 Cap: 10000 Floor: 20 Init: -10000 P: undead	compounded experience (c100)	time residue (c0)
1914 Apathetic Tyrannosaurus	1959	apatheticdino.gif	Scale: 3 Cap: ? Floor: 20 Init: -10000 P: beast	compounded experience (c100)	time residue (c0)
1914 Apathetic Tyrannosaurus	1959	apatheticdino.gif	Scale: 3 Cap: 10000 Floor: 20 Init: -10000 P: beast	compounded experience (c100)	time residue (c0)
1915 Assembly Elemental	1960	assemblyelem.gif	Scale: 5 Cap: ? Floor: 20 Init: -10000 P: elemental	compounded experience (c100)	time residue (c0)
1915 Assembly Elemental	1960	assemblyelem.gif	Scale: 5 Cap: 10000 Floor: 20 Init: -10000 P: elemental	compounded experience (c100)	time residue (c0)
1916 Cro-Magnon Gnoll	1961	cromaggnoll.gif	Scale: 7 Cap: ? Floor: 20 Init: -10000 P: humanoid	compounded experience (c100)	time residue (c0)
1916 Cro-Magnon Gnoll	1961	cromaggnoll.gif	Scale: 7 Cap: 10000 Floor: 20 Init: -10000 P: humanoid	compounded experience (c100)	time residue (c0)
1917 Krakrox the Barbarian	1964	krakrox.gif	Scale: 25 Cap: ? Floor: 40 Init: -10000 P: dude	compounded experience (c100)	time residue (c100)	time residue (c100)	time residue (c100)
1917 Krakrox the Barbarian	1964	krakrox.gif	Scale: 25 Cap: 10000 Floor: 40 Init: -10000 P: dude	compounded experience (c100)	time residue (c100)	time residue (c100)	time residue (c100)
1919 Wooly Duck	1963	woolyduck.gif	Scale: 5 Cap: ? Floor: 20 Init: -10000 P: beast	compounded experience (c100)	time residue (c0)
1919 Wooly Duck	1963	woolyduck.gif	Scale: 5 Cap: 10000 Floor: 20 Init: -10000 P: beast	compounded experience (c100)	time residue (c0)
1939 GNG-3-R	1992	gng3r.gif	NOCOPY BOSS Scale: ? Cap: ? Floor: ? Init: 300 P: construct	industrial frosting (n100)
1939 GNG-3-R	1992	gng3r.gif	BOSS NOCOPY Scale: 0 Cap: 10000 Floor: 10 Init: 300 P: construct	industrial frosting (n100)
1942 LOV Enforcer	2009	lovenforcer.gif	NOCOPY Scale: 5 Cap: ? Floor: ? Init: -10000 P: dude Phys: 75	LOV Elixir #3 (c0)
1942 LOV Enforcer	2009	lovenforcer.gif	NOCOPY Scale: 5 Cap: 10000 Floor: 10 Init: -10000 P: dude Phys: 75	LOV Elixir #3 (c0)
1943 LOV Engineer	2010	lovengineer.gif	NOCOPY Scale: 5 Cap: ? Floor: ? Init: -10000 P: dude Phys: 25	LOV Elixir #6 (c0)
1943 LOV Engineer	2010	lovengineer.gif	NOCOPY Scale: 5 Cap: 10000 Floor: 10 Init: -10000 P: dude Phys: 25	LOV Elixir #6 (c0)
1944 LOV Equivocator	2011	lovequivocator.gif	NOCOPY Scale: 5 Cap: ? Floor: ? Init: 500 P: dude Phys: 50	LOV Elixir #9 (p0)
1944 LOV Equivocator	2011	lovequivocator.gif	NOCOPY Scale: 5 Cap: 10000 Floor: 10 Init: 500 P: dude Phys: 50	LOV Elixir #9 (p0)
1947 hostile plant	2013	sgplanta1.gif,sgplanta2.gif,sgplanta3.gif,sgplanta4.gif,sgplanta5.gif,sgplanta6.gif,sgplanta7.gif,sgplanta8.gif,sgplanta9.gif,sgplanta10.gif,sgplanta11.gif,sgplanta12.gif,sgplanta13.gif,sgplanta14.gif,sgplanta15.gif,sgplanta16.gif,sgplanta17.gif,sgplanta18.gif,sgplanta19.gif,sgplanta20.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: ? Init: [60+10*pref(_spacegatePlanetIndex)] P: plant	edible alien plant bit	alien plant fibers	alien plant goo
1947 hostile plant	2013	sgplanta1.gif,sgplanta2.gif,sgplanta3.gif,sgplanta4.gif,sgplanta5.gif,sgplanta6.gif,sgplanta7.gif,sgplanta8.gif,sgplanta9.gif,sgplanta10.gif,sgplanta11.gif,sgplanta12.gif,sgplanta13.gif,sgplanta14.gif,sgplanta15.gif,sgplanta16.gif,sgplanta17.gif,sgplanta18.gif,sgplanta19.gif,sgplanta20.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: 10 Init: [60+10*pref(_spacegatePlanetIndex)] P: plant	edible alien plant bit	alien plant fibers	alien plant goo
1948 large hostile plant	2014	sgplantb1.gif,sgplantb2.gif,sgplantb3.gif,sgplantb4.gif,sgplantb5.gif,sgplantb6.gif,sgplantb7.gif,sgplantb8.gif,sgplantb9.gif,sgplantb10.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: ? Init: [60+10*pref(_spacegatePlanetIndex)] Init: -10000 P: plant	edible alien plant bit	edible alien plant bit	alien plant fibers	alien plant fibers	alien plant goo
1948 large hostile plant	2014	sgplantb1.gif,sgplantb2.gif,sgplantb3.gif,sgplantb4.gif,sgplantb5.gif,sgplantb6.gif,sgplantb7.gif,sgplantb8.gif,sgplantb9.gif,sgplantb10.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: 10 Init: -10000 P: plant	edible alien plant bit	edible alien plant bit	alien plant fibers	alien plant fibers	alien plant goo
1949 exotic hostile plant	2015	sgplantc1.gif,sgplantc2.gif,sgplantc3.gif	Scale: [10+25*pref(_spacegatePlanetIndex)] NOCOPY Cap: 10000 Floor: ? Init: [60+10*pref(_spacegatePlanetIndex)] P: plant	edible alien plant bit	edible alien plant bit	edible alien plant bit	alien plant fibers	alien plant fibers	alien plant fibers	alien plant goo	alien plant pod
1949 exotic hostile plant	2015	sgplantc1.gif,sgplantc2.gif,sgplantc3.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: 10 Init: [60+10*pref(_spacegatePlanetIndex)] P: plant	edible alien plant bit	edible alien plant bit	edible alien plant bit	alien plant fibers	alien plant fibers	alien plant fibers	alien plant goo	alien plant pod
1950 small hostile animal	2016	sganimala1.gif,sganimala2.gif,sganimala3.gif,sganimala4.gif,sganimala5.gif,sganimala6.gif,sganimala7.gif,sganimala8.gif,sganimala9.gif,sganimala10.gif,sganimala11.gif,sganimala12.gif,sganimala13.gif,sganimala14.gif,sganimala15.gif,sganimala16.gif,sganimala17.gif,sganimala18.gif,sganimala19.gif,sganimala20.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: ? Init: [60+10*pref(_spacegatePlanetIndex)] P: beast	alien meat	alien toenails	alien animal goo
1950 small hostile animal	2016	sganimala1.gif,sganimala2.gif,sganimala3.gif,sganimala4.gif,sganimala5.gif,sganimala6.gif,sganimala7.gif,sganimala8.gif,sganimala9.gif,sganimala10.gif,sganimala11.gif,sganimala12.gif,sganimala13.gif,sganimala14.gif,sganimala15.gif,sganimala16.gif,sganimala17.gif,sganimala18.gif,sganimala19.gif,sganimala20.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: 10 Init: [60+10*pref(_spacegatePlanetIndex)] P: beast	alien meat	alien toenails	alien animal goo
1951 large hostile animal	2017	sganimalb1.gif,sganimalb2.gif,sganimalb3.gif,sganimalb4.gif,sganimalb5.gif,sganimalb6.gif,sganimalb7.gif,sganimalb8.gif,sganimalb9.gif,sganimalb10.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: ? Init: [60+10*pref(_spacegatePlanetIndex)] P: beast	alien meat	alien meat	alien toenails	alien toenails	alien animal goo
1951 large hostile animal	2017	sganimalb1.gif,sganimalb2.gif,sganimalb3.gif,sganimalb4.gif,sganimalb5.gif,sganimalb6.gif,sganimalb7.gif,sganimalb8.gif,sganimalb9.gif,sganimalb10.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: 10 Init: [60+10*pref(_spacegatePlanetIndex)] P: beast	alien meat	alien meat	alien toenails	alien toenails	alien animal goo
1952 exotic hostile animal	2018	sganimalc1.gif,sganimalc2.gif,sganimalc3.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: ? Init: [60+10*pref(_spacegatePlanetIndex)] P: beast	alien meat	alien meat	alien meat	alien toenails	alien toenails	alien toenails	alien animal goo	alien animal milk
1952 exotic hostile animal	2018	sganimalc1.gif,sganimalc2.gif,sganimalc3.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: 10 Init: [60+10*pref(_spacegatePlanetIndex)] P: beast	alien meat	alien meat	alien meat	alien toenails	alien toenails	alien toenails	alien animal goo	alien animal milk
1953 Spant drone	2019	sgspantdrone.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: ? Init: [60+10*pref(_spacegatePlanetIndex)] P: bug	spant chitin	spant chitin	spant tendon	spant tendon
1953 Spant drone	2019	sgspantdrone.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: 10 Init: [60+10*pref(_spacegatePlanetIndex)] P: bug	spant chitin	spant chitin	spant tendon	spant tendon
1954 Spant soldier	2020	sgspantwarrior.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: ? Init: [60+10*pref(_spacegatePlanetIndex)] P: bug	spant chitin	spant chitin	spant tendon	spant tendon	spant spear
1954 Spant soldier	2020	sgspantwarrior.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: 10 Init: [60+10*pref(_spacegatePlanetIndex)] P: bug	spant chitin	spant chitin	spant tendon	spant tendon	spant spear
1955 Murderbot drone	2021	sgmbdrone.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: ? Init: [60+10*pref(_spacegatePlanetIndex)] P: construct	murderbot component casing	murderbot monofilament	murderbot monofilament	murderbot power cell	murderbot power cell	murderbot memory chip
1955 Murderbot drone	2021	sgmbdrone.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: 10 Init: [60+10*pref(_spacegatePlanetIndex)] P: construct	murderbot component casing	murderbot monofilament	murderbot monofilament	murderbot power cell	murderbot power cell	murderbot memory chip
1956 Murderbot soldier	2022	sgmb.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: ? Init: [60+10*pref(_spacegatePlanetIndex)] P: construct	murderbot component casing	murderbot component casing	murderbot monofilament	murderbot monofilament	murderbot power cell	murderbot power cell	murderbot memory chip	murderbot memory chip	murderbot memory chip	murderbot plasma rifle
1956 Murderbot soldier	2022	sgmb.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: 10 Init: [60+10*pref(_spacegatePlanetIndex)] P: construct	murderbot component casing	murderbot component casing	murderbot monofilament	murderbot monofilament	murderbot power cell	murderbot power cell	murderbot memory chip	murderbot memory chip	murderbot memory chip	murderbot plasma rifle
1957 hostile intelligent alien	2023	sgalienb1.gif,sgalienb2.gif,sgalienb3.gif,sgalienb4.gif,sgalienb5.gif,sgalienb6.gif,sgalienb7.gif,sgalienb8.gif,sgalienb9.gif,sgalienb10.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: ? Init: [60+10*pref(_spacegatePlanetIndex)] P: humanoid
1957 hostile intelligent alien	2023	sgalienb1.gif,sgalienb2.gif,sgalienb3.gif,sgalienb4.gif,sgalienb5.gif,sgalienb6.gif,sgalienb7.gif,sgalienb8.gif,sgalienb9.gif,sgalienb10.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: 10 Init: [60+10*pref(_spacegatePlanetIndex)] P: humanoid
1998 God Lobster	2088	godlob.gif,godlob_scepter.gif,godlob_ring.gif,godlob_rod.gif,godlob_cape.gif,godlob_crown.gif	NOCOPY FREE Scale: [10*equipped(God Lobster's Scepter)+20*equipped(God Lobster's Ring)+30*equipped(God Lobster's Rod)+40*equipped(God Lobster's Robe)+50*equipped(God Lobster's Crown)] Cap: 1500 Floor: ? Init: 1000 P: horror
1998 God Lobster	2088	godlob.gif,godlob_scepter.gif,godlob_ring.gif,godlob_rod.gif,godlob_cape.gif,godlob_crown.gif	NOCOPY FREE Scale: [10*equipped(God Lobster's Scepter)+20*equipped(God Lobster's Ring)+30*equipped(God Lobster's Rod)+40*equipped(God Lobster's Robe)+50*equipped(God Lobster's Crown)] Cap: 1500 Floor: 10 Init: 1000 P: horror
2015 sausage goblin	2104	sausagegoblin.gif	FREE Scale: [1+2*pref(_sausageFights)] Cap: ? Floor: ? Init: -10000 Meat: 69 E: sleaze P: goblin	magical sausage casing (n100)	red-hot sausage fork (0)	bag of sausage links (n0)
2015 sausage goblin	2104	sausagegoblin.gif	FREE Scale: [1+2*pref(_sausageFights)] Cap: 10000 Floor: 10 Init: -10000 Meat: 69 P: goblin E: sleaze	magical sausage casing (n100)	red-hot sausage fork (0)	bag of sausage links (n0)
2051 holographic army	1629	holoarmy.gif	NOCOPY Scale: 30 Floor: ? Init: -10000 P: weird	holo-bomber (0)	holo-platoon (0)	holo-tank (0)
2051 holographic army	1629	holoarmy.gif	NOCOPY Scale: 30 Floor: 10 Init: -10000 P: weird	holo-bomber (0)	holo-platoon (0)	holo-tank (0)
2053 Xiblaxian political prisoner	1630	polprisoner.gif	NOCOPY Scale: 40 Floor: ? Init: -10000 P: weird	residual zeal (n100)	little red .epub file (c0)
2053 Xiblaxian political prisoner	1630	polprisoner.gif	NOCOPY Scale: 40 Floor: 10 Init: -10000 P: weird	residual zeal (n100)	little red .epub file (c0)
2109 Drunken Tropical Vacationer	1772	kok_drunk.gif	Scale: 0 Cap: ? Init: -10000 P: dude Meat: 200	Kokomo Resort Brand Suntan Oil (0)	Kokomo Resort Chip (0)	Kokomo Resort Chip (0)	Kokomo Resort Chip (0)	captured boozles (c0)
2109 Drunken Tropical Vacationer	1772	kok_drunk.gif	Scale: 0 Cap: 10000 Init: -10000 Meat: 200 P: dude	Kokomo Resort Brand Suntan Oil (0)	Kokomo Resort Chip (0)	Kokomo Resort Chip (0)	Kokomo Resort Chip (0)	captured boozles (c0)
2110 Lovestruck Tropical Honeymooners	1771	kok_lovers.gif	Scale: -5 Cap: ? Init: -10000 P: dude Meat: 250	Afternoon Delight (0)	Kokomo Resort Chip (0)	Kokomo Resort Chip (0)
2110 Lovestruck Tropical Honeymooners	1771	kok_lovers.gif	Scale: -5 Cap: 10000 Init: -10000 Meat: 250 P: dude	Afternoon Delight (0)	Kokomo Resort Chip (0)	Kokomo Resort Chip (0)
2111 Resort Waiter	1773	kok_waiter.gif	Scale: 5 Cap: ? Init: -10000 P: dude Meat: 100	Kokomo Resort Order Pad (0)	Kokomo Resort Chip (0)	Kokomo Resort Chip (0)	drinks tray (c0)
2111 Resort Waiter	1773	kok_waiter.gif	Scale: 5 Cap: 10000 Init: -10000 Meat: 100 P: dude	Kokomo Resort Order Pad (0)	Kokomo Resort Chip (0)	Kokomo Resort Chip (0)	drinks tray (c0)
2112 Brick Mulligan, the Bartender	1774	kok_tender.gif	NOCOPY Scale: 5 Cap: ? Floor: ? Init: -10000 P: dude	lemon (0)	lime (0)	orange (0)	bottle of gin (0)	bottle of vodka (0)	bottle of rum (0)	coconut shell (0)	little paper umbrella (0)	magical ice cubes (0)	eyebrow lifter (c0)
2112 Brick Mulligan, the Bartender	1774	kok_tender.gif	NOCOPY Scale: 5 Cap: 10000 Floor: 10 Init: -10000 P: dude	lemon (0)	lime (0)	orange (0)	bottle of gin (0)	bottle of vodka (0)	bottle of rum (0)	coconut shell (0)	little paper umbrella (0)	magical ice cubes (0)	eyebrow lifter (c0)
2334 mutant circuit-soldering elf	751	antlerelf.gif	Scale: 8 Floor: ? Init: 150 P: elf	elven moonshine (0)	miniature antlers (0)
2334 mutant circuit-soldering elf	751	antlerelf.gif	Scale: 8 Floor: 10 Init: 150 P: elf	elven moonshine (0)	miniature antlers (0)
2335 mutant cookie-baking elf	749	elfblob.gif	Scale: 6 Floor: ? Init: 75 P: elf	festive holiday hat (0)
2335 mutant cookie-baking elf	749	elfblob.gif	Scale: 6 Floor: 10 Init: 75 P: elf	festive holiday hat (0)
2336 mutant doll-dressing elf	748	elflimbs.gif	Scale: 6 Floor: ? Init: 75 P: elf	cheap elven gloves (0)
2336 mutant doll-dressing elf	748	elflimbs.gif	Scale: 6 Floor: 10 Init: 75 P: elf	cheap elven gloves (0)
2337 mutant gift-wrapping elf	747	elfclaw.gif	Scale: 6 Floor: ? Init: 75 P: elf	elven socks (0)
2337 mutant gift-wrapping elf	747	elfclaw.gif	Scale: 6 Floor: 10 Init: 75 P: elf	elven socks (0)
2338 mutant whistle-carving elf	750	elfhulk.gif	Scale: 7 Floor: ? Init: 100 P: elf	elven <i>limbos</i> gingerbread (0)	elven whittling knife (0)
2338 mutant whistle-carving elf	750	elfhulk.gif	Scale: 7 Floor: 10 Init: 100 P: elf	elven <i>limbos</i> gingerbread (0)	elven whittling knife (0)
2348 deadwood tree	864	shiv_dead.gif	Scale: 5 Cap: 200 Floor: ? Init: -10000 P: plant	petrified wood (5)	spooky bark (10)	spooky sap (5)
2348 deadwood tree	864	shiv_dead.gif	Scale: 5 Cap: 200 Floor: 10 Init: -10000 P: plant	petrified wood (5)	spooky bark (10)	spooky sap (5)
2349 fur tree	862	shiv_fur.gif	Scale: 5 Cap: 200 Floor: ? Init: -10000 P: plant	petrified wood (5)	spooky bark (10)	spooky sap (5)	wolfman mask (20)
2349 fur tree	862	shiv_fur.gif	Scale: 5 Cap: 200 Floor: 10 Init: -10000 P: plant	petrified wood (5)	spooky bark (10)	spooky sap (5)	wolfman mask (20)
2350 hangman's tree	865	shiv_hangman.gif	Scale: 5 Cap: 200 Floor: ? Init: -10000 P: plant	petrified wood (5)	spooky bark (10)	spooky sap (5)
2350 hangman's tree	865	shiv_hangman.gif	Scale: 5 Cap: 200 Floor: 10 Init: -10000 P: plant	petrified wood (5)	spooky bark (10)	spooky sap (5)
2351 pumpkin tree	863	shiv_pumpkin.gif	Scale: 5 Cap: 200 Floor: ? Init: -10000 P: plant	petrified wood (5)	pumpkinhead mask (20)	spooky bark (10)	spooky sap (5)
2351 pumpkin tree	863	shiv_pumpkin.gif	Scale: 5 Cap: 200 Floor: 10 Init: -10000 P: plant	petrified wood (5)	pumpkinhead mask (20)	spooky bark (10)	spooky sap (5)
2352 toilet-papered tree	861	shiv_tp.gif	Scale: 5 Cap: 200 Floor: ? Init: -10000 P: plant	mummy costume (20)	petrified wood (5)	spooky bark (10)	spooky sap (5)
2352 toilet-papered tree	861	shiv_tp.gif	Scale: 5 Cap: 200 Floor: 10 Init: -10000 P: plant	mummy costume (20)	petrified wood (5)	spooky bark (10)	spooky sap (5)
2358 bolt-cuttin' elf	900	elf_boltcutters.gif	Scale: ? Cap: ? Floor: ? Init: -10000 P: elf	elf resistance button (100)	handful of headless bolts (c100)	pair of bolt cutters (0)
2358 bolt-cuttin' elf	900	elf_boltcutters.gif	Scale: 0 Cap: 10000 Floor: 10 Init: -10000 P: elf	elf resistance button (100)	handful of headless bolts (c100)	pair of bolt cutters (0)
2359 monkey wrenchin' elf	894	elf_wrench.gif	Scale: ? Cap: ? Floor: ? Init: -10000 P: elf	elf resistance button (100)	throwing wrench (0)	wrench handle (c100)
2359 monkey wrenchin' elf	894	elf_wrench.gif	Scale: 0 Cap: 10000 Floor: 10 Init: -10000 P: elf	elf resistance button (100)	throwing wrench (0)	wrench handle (c100)
2360 propaganda-spewin' elf	901	elf_propaganda.gif	Scale: ? Cap: ? Floor: ? Init: -10000 P: elf	bottle of agitprop ink (c100)	elf resistance button (100)	poison pen (0)
2360 propaganda-spewin' elf	901	elf_propaganda.gif	Scale: 0 Cap: 10000 Floor: 10 Init: -10000 P: elf	bottle of agitprop ink (c100)	elf resistance button (100)	poison pen (0)
2423 Hammered Yam Golem	-57	drunkyam.gif	NOCOPY WANDERER NOMANUEL  Scale: -3 MLMult: 0	marshmallow flamb&eacute; (c100)
2423 Hammered Yam Golem	-57	drunkyam.gif	NOCOPY NOMANUEL WANDERER Scale: -3 MLMult: 0	marshmallow flamb&eacute; (c100)
2497 depressing French accordionist	1452	wanderacc1.gif	WANDERER Scale: -3 Floor: ? Init: -10000 P: dude	Bal-musette accordion (a0)
2497 depressing French accordionist	1452	wanderacc1.gif	WANDERER Scale: -3 Floor: 10 Init: -10000 P: dude	Bal-musette accordion (a0)
2498 lively Cajun accordionist	1453	wanderacc2.gif	WANDERER Scale: -3 Floor: ? Init: -10000 P: dude	Cajun accordion (a0)
2498 lively Cajun accordionist	1453	wanderacc2.gif	WANDERER Scale: -3 Floor: 10 Init: -10000 P: dude	Cajun accordion (a0)
2499 quirky indie-rock accordionist	1454	wanderacc3.gif	WANDERER Scale: -3 Floor: ? Init: -10000 P: dude	quirky accordion (a0)
2499 quirky indie-rock accordionist	1454	wanderacc3.gif	WANDERER Scale: -3 Floor: 10 Init: -10000 P: dude	quirky accordion (a0)
2518 Warbear Officer	1467	warbear21.gif,warbear22.gif,warbear23.gif	Scale: ? Cap: ? Floor: ? Init: 50 P: beast	warbear whosit (n100)	warbear whosit (n100)	warbear requisition box (f0)	warbear badge (f0)
2518 Warbear Officer	1467	warbear21.gif,warbear22.gif,warbear23.gif	Scale: 0 Cap: 10000 Floor: 10 Init: 50 P: beast	warbear whosit (n100)	warbear whosit (n100)	warbear requisition box (f0)	warbear badge (f0)
2519 High-Ranking Warbear Officer	1468	warbear31.gif,warbear32.gif,warbear33.gif	Scale: ? Cap: ? Floor: ? Init: 100 P: beast	warbear whosit (n100)	warbear whosit (n100)	warbear whosit (n100)	warbear officer requisition box (f0)
2519 High-Ranking Warbear Officer	1468	warbear31.gif,warbear32.gif,warbear33.gif	Scale: 0 Cap: 10000 Floor: 10 Init: 100 P: beast	warbear whosit (n100)	warbear whosit (n100)	warbear whosit (n100)	warbear officer requisition box (f0)
2552 imagining how your life would be better if you'd made different decisions	1995	badvibe3.gif	Scale: +3 Cap: 10000 Floor: 20 Init: -10000 P: weird	negative lump (0)	chakra sludge (0)
2552 imagining how your life would be better if you'd made different decisions	1995	badvibe3.gif	Scale: 3 Cap: 10000 Floor: 20 Init: -10000 P: weird	negative lump (0)	chakra sludge (0)
```

midgleyc commented:
I note that:

```
Scale: ? Cap: ? Floor: ?
```

is replaced with 

```
Scale: 0 Cap: 10000 Floor: 10
```

Not sure if deliberate or correct.

`Gnollish Crossdresser` had both an `E: sleaze` and an `EA: sleaze`. Probably an error.

Some spaces were collapsed -- good.

Veracity0 commented:
> I note that:
> 
> ```
> Scale: ? Cap: ? Floor: ?
> ```
> 
> is replaced with
> 
> ```
> Scale: 0 Cap: 10000 Floor: 10
> ```
> 
> Not sure if deliberate or correct.

It is deliberate for the MonsterData to have an int for those -- and it intentionally uses the default if the value is "?" - unknown.

Yet another place where I should defer evaluation (and defaulting) of "?" to an int in the constructor.

> `Gnollish Crossdresser` had both an `E: sleaze` and an `EA: sleaze`. Probably an error.

Yes. Its element is sleaze and it has a sleaze attack. It should just be E:

jaadams5 commented:
DataFileMechanicsTest just checks the number of delimited fields in a line, for various data files.  The previous iteration of monsters.txt was too complex for that.  Once this work is done we might consider whether the new version is simpler and thus a candidate for that test.  I'll monitor but this will help remind me or start a sidebar discussion about whether DataFileMechanicsTest has any utility as a test.

Veracity0 commented:
I stuck in some logging:


```
> test fix-monster-attributes

Monster: "Gnollish Crossdresser" has both EA: sleaze and sleaze
Monster: "Knott Slanding" has meat range 970-1342
Monster: "lihc" has both EA: spooky and sleaze
919 monster attributes changed
```
Looks like you found all the "weird" monsters. :)

Veracity0 commented:
I fixed Scale, Cap, and Floor to allow either an int, an expression, or ? - which means unknown so use default.
Those are all calculated at monster expression time. But the attribute has the original form.
I added a test for those three factors with "?".

I also reran the "fix" script on the original monsters.txt and included the fixed file.

Take another look, please?

Veracity0 commented:
> DataFileMechanicsTest just checks the number of delimited fields in a line, for various data files. The previous iteration of monsters.txt was too complex for that. Once this work is done we might consider whether the new version is simpler and thus a candidate for that test. I'll monitor but this will help remind me or start a sidebar discussion about whether DataFileMechanicsTest has any utility as a test.

The new version is not "simpler" in that the number of delimited fields did not change; there are 4 required fields and then an arbitrary number of tab-separated item drops.

Now, if DataFileMechanicsTest wants to test the CONTENTS of the various fields, it could certainly validate whether the "attributes" field of this file are correct. Of course, running KoLmafia will do that for you. :)

midgleyc reviewed (changes requested):
`large hostile plant` had `Init` present twice, and has been collapsed to once. This was probably a mistake? 

Current differing lines (again excluding posion):

```
292 Gnollish Crossdresser	19	dk_cross.gif	Atk: 11 Def: 9 HP: 5 Init: 60 E: sleaze EA: sleaze Meat: 22 P: humanoid	gnoll teeth (5)	frilly skirt (25)	maiden wig (25)	frilly shirt (c5)	Gnollish Crossdress (c0)
292 Gnollish Crossdresser	19	dk_cross.gif	Atk: 11 Def: 9 HP: 5 Init: 60 Meat: 22 P: humanoid E: sleaze	gnoll teeth (5)	frilly skirt (25)	maiden wig (25)	frilly shirt (c5)	Gnollish Crossdress (c0)
570 scary pirate	307	zompirate.gif	Atk: 150 Def: 135 HP: 200 Init: 10000 E: spooky Meat: 45 P: pirate Wiki: "scary pirate (cursed)" 	cursed breeches (n0)	cursed cutlass (n0)	cursed eyepatch (n0)	folder (skull and crossbones) (c0)
570 scary pirate	307	zompirate.gif	Atk: 150 Def: 135 HP: 200 Init: 10000 Meat: 45 P: pirate E: spooky Wiki: "scary pirate (cursed)"	cursed breeches (n0)	cursed cutlass (n0)	cursed eyepatch (n0)	folder (skull and crossbones) (c0)
638 Stephen Spookyraven	1565	steven.gif	NOCOPY Scale: 5 Cap: 300 Init: 100 Meat: 500  P: undead	Stephen's secret formula (100)	Stephen's lab coat (100)
638 Stephen Spookyraven	1565	steven.gif	NOCOPY Scale: 5 Cap: 300 Init: 100 Meat: 500 P: undead	Stephen's secret formula (100)	Stephen's lab coat (100)
681 The Thing in the Basement	2164	drippything1.gif	NOMANUEL Atk: [300] Def: [300] HP: [320] Init: -10000 P: horror DRIPPY 	The Eye of the Thing in the Basement (n100)	The Fingernail of the Thing in the Basement (n100)
681 The Thing in the Basement	2164	drippything1.gif	NOMANUEL Atk: [300] Def: [300] HP: [320] Init: -10000 P: horror DRIPPY	The Eye of the Thing in the Basement (n100)	The Fingernail of the Thing in the Basement (n100)
1948 large hostile plant	2014	sgplantb1.gif,sgplantb2.gif,sgplantb3.gif,sgplantb4.gif,sgplantb5.gif,sgplantb6.gif,sgplantb7.gif,sgplantb8.gif,sgplantb9.gif,sgplantb10.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: ? Init: [60+10*pref(_spacegatePlanetIndex)] Init: -10000 P: plant	edible alien plant bit	edible alien plant bit	alien plant fibers	alien plant fibers	alien plant goo
1948 large hostile plant	2014	sgplantb1.gif,sgplantb2.gif,sgplantb3.gif,sgplantb4.gif,sgplantb5.gif,sgplantb6.gif,sgplantb7.gif,sgplantb8.gif,sgplantb9.gif,sgplantb10.gif	NOCOPY Scale: [10+25*pref(_spacegatePlanetIndex)] Cap: 10000 Floor: ? Init: -10000 P: plant	edible alien plant bit	edible alien plant bit	alien plant fibers	alien plant fibers	alien plant goo
2052 They	1631	they.gif	Scale: 20  Floor: ? Init: -10000 P: humanoid	They liver (0)	They liver (0)	BUBBLE GUM (c0)
2052 They	1631	they.gif	Scale: 20 Floor: ? Init: -10000 P: humanoid	They liver (0)	They liver (0)	BUBBLE GUM (c0)
2423 Hammered Yam Golem	-57	drunkyam.gif	NOCOPY WANDERER NOMANUEL  Scale: -3 MLMult: 0	marshmallow flamb&eacute; (c100)
2423 Hammered Yam Golem	-57	drunkyam.gif	NOCOPY NOMANUEL WANDERER Scale: -3 MLMult: 0	marshmallow flamb&eacute; (c100)
2552 imagining how your life would be better if you'd made different decisions	1995	badvibe3.gif	Scale: +3 Cap: 10000 Floor: 20 Init: -10000 P: weird	negative lump (0)	chakra sludge (0)
2552 imagining how your life would be better if you'd made different decisions	1995	badvibe3.gif	Scale: 3 Cap: 10000 Floor: 20 Init: -10000 P: weird	negative lump (0)	chakra sludge (0)
```

* collapse of identical EA and E: Crossdresser 
* collapse of whitespace: scary pirate, Stephen Spookyraven, The Thing in the Basement, They, Hammered Yam Golem
* Scale: +3 to 3: imagining how your life would be better if you'd made different decisions
* collapse of double init: large hostile plant

Other than the plant, this all seems sensible.

Veracity0 commented:
I moved the duplicate attribute check up in front of the switch statement which handles them.
```
Monster: "lihc" has multiple values for the "EA:" attribute.
Monster: "large hostile plant" has multiple values for the "Init:" attribute.
```
This means that I can't be specific about the two values (since we haven't parsed them yet), but it would catch the error with the two Init: values.

If we add support for multiple EA:, this will have to change - just as the code for parsing EA: will have to change. :)

I fixed the large hostile plant to have the formula.

Veracity0 commented:
[midgleyc](https://github.com/midgleyc) requested changes on behalf of https://github.com/orgs/kolmafia/teams/staff [4 hours ago](https://github.com/kolmafia/kolmafia/pull/793#pullrequestreview-995944392)

> `large hostile plant` had `Init` present twice (one formula, one number), and has been collapsed to one (number). This was probably a mistake in the initial data?
> 
> Other than the plant, this all seems sensible.

I fixed that, but I don't seem to have the option to "resolve" the request?
